providers/google_test.go

```diff
@@ -109,21 +110,50 @@ func TestGoogleProviderGetEmailAddress(t *testing.T) {
 	assert.Equal(t, "refresh12345", session.RefreshToken)
 }
 
-func TestGoogleProviderValidateGroup(t *testing.T) {
-	p := newGoogleProvider()
-	p.GroupValidator = func(email string) bool {
-		return email == "michael.bland@gsa.gov"
+func TestGoogleProviderGroupValidator(t *testing.T) {
+	const sessionEmail = "michael.bland@gsa.gov"
+
+	testCases := map[string]struct {
+		session       *sessions.SessionState
+		validatorFunc func(*sessions.SessionState) bool
+		expectedAuthZ bool
+	}{
+		"Email is authorized with GroupValidator": {
+			session: &sessions.SessionState{
+				Email: sessionEmail,
+			},
+			validatorFunc: func(s *sessions.SessionState) bool {
+				return s.Email == sessionEmail
+			},
+			expectedAuthZ: true,
+		},
+		"Email is denied with GroupValidator": {
+			session: &sessions.SessionState{
+				Email: sessionEmail,
+			},
+			validatorFunc: func(s *sessions.SessionState) bool {
+				return s.Email != sessionEmail
+			},
+			expectedAuthZ: false,
+		},
+		"Default does no authorization checks": {
+			session: &sessions.SessionState{
+				Email: sessionEmail,
+			},
+			validatorFunc: nil,
+			expectedAuthZ: true,
+		},
 	}
-	assert.Equal(t, true, p.ValidateGroup("michael.bland@gsa.gov"))
-	p.GroupValidator = func(email string) bool {
-		return email != "michael.bland@gsa.gov"
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			g := NewWithT(t)
+			p := newGoogleProvider()
+			if tc.validatorFunc != nil {
+				p.GroupValidator = tc.validatorFunc
+			}
+			g.Expect(p.GroupValidator(tc.session)).To(Equal(tc.expectedAuthZ))
```

Comment:
Does this test actually test the code or does it just test the testCase validator func? 🤔

Reply:
Tests the code's usage of an assigned `GroupValidator` and the code's default validator when no groups are configured (the third test case, allow all).

But you are right, no test of the new logic in the configured `GroupValidator` setup in `SetGroupRestriction`.

It used to just call `userInGroup` (which has its own tests).  The added logic just takes those groups and appends them to `session.Groups`.

I agree though, in an ideal world, we should have a test for that (I'm not a fan in general of this Google validatorFunc handler obfuscation) -- but I think I got scared of needing to mock/refactor `getAdminService` and the `credentialsReader io.Reader` aspects.

I'll add a TODO comment to unit test or refactor for a separate PR.